opm/core/props/BlackoilPhases.cpp

```diff
@@ -0,0 +1,51 @@
+/*
+  Copyright 2021 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <vector>
+
+#include <opm/core/props/BlackoilPhases.hpp>
+
+Opm::PhaseUsage::PhaseUsage(const std::vector<BlackoilPhases::PhaseIndex>& phases)
+{
+    this->phase_used.fill(0);
+    this->phase_pos.fill(-1);
+
+    std::size_t current_pos = 0;
+    for (const auto& phase : phases) {
+        this->phase_used[phase] = 1;
+        this->phase_pos[phase] = current_pos;
```

Comment:
I believe this breaks an assumption: that the ordering of active "phases" is the same as the canonical ordering (as given by the line
```
enum PhaseIndex { Aqua = 0, Liquid = 1, Vapour = 2, Solvent = 3, Polymer = 4, Energy = 5, PolymerMW = 6, Foam = 7, Brine = 8, ZFraction = 9 };
```
Indeed the test explicitly test this by passing Polymer before Solvent. I think the resulting positions should *not* depend on (and be the same as) the input order, but be the same irrespective of input order, the test is therefore surprising to me.

Now, I cannot immediately come up with some code lines that break if this assumption is dropped. But just the fact that "this is surprising" makes me worry that there could be some, and I would prefer to avoid having to hunt for them...

Otherwise this looks good and should be merged if we can reach a consensus on the ordering issue.

Reply:
I do not know these details - I have tried to create a constructor based on my understanding of the `createFromDeck` function; I obviously missed a point here. Currently the usage order will follow the (random and arbitrary) input order. I will update the PR to ensure that the items come in the canonical order.

Reply:
Have added a `std::sort()` to ensure that canonical ordering is respected. Have also removed an unused utility class `PhasePresence`